### PR TITLE
Fix NRE in linq processing of custom components

### LINKED ISSUE #2937: NullReferenceException in version 5.3.10
With the latest version (5.3.10) of NHibernate I receive a NullReferenceException.
Version 5.3.9 works

.NET 6 RC2

Query:
```c#
DataManager.Query<WeatherDataType>(transaction).FirstOrDefault(t => t.Identification.SensorCode == sensorCode)
```

```c#
[Property(0, TypeType = typeof(WeatherDataIdentificationCompositeUserType))]
[Column(1, Name = nameof(WeatherDataIdentification.DateOfSampling), Unique = false, NotNull = false, Index = "WEATHER_DATEOFSAMPLING")]
[Column(2, Name = nameof(WeatherDataIdentification.InternalId), Unique = false, NotNull = false)]
[Column(3, Name = nameof(WeatherDataIdentification.Position) + "_Latitude", Unique = false, NotNull = false)]
[Column(4, Name = nameof(WeatherDataIdentification.Position) + "_Longitude", Unique = false, NotNull = false)]
[Column(5, Name = nameof(WeatherDataIdentification.SensorCode), Unique = false, NotNull = false)]
[Column(6, Name = nameof(WeatherDataIdentification.SensorId), Unique = false, NotNull = false)]
public virtual WeatherDataIdentification Identification...
```

StackTrace:
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at NHibernate.Util.ExpressionsHelper.TryGetMappedNullability(ISessionFactoryImplementor sessionFactory, Expression expression, Boolean& nullable)
   at NHibernate.Linq.Visitors.NullableExpressionDetector.IsNullable(MemberExpression memberExpression, BinaryExpression equalityExpression)
   at NHibernate.Linq.Visitors.NullableExpressionDetector.IsNullable(Expression expression, BinaryExpression equalityExpression)
   at NHibernate.Linq.Visitors.HqlGeneratorExpressionVisitor.TranslateEqualityComparison(BinaryExpression expression)
   at NHibernate.Linq.Visitors.HqlGeneratorExpressionVisitor.VisitExpression(Expression expression)
   at NHibernate.Linq.Visitors.QueryModelVisitor.VisitWhereClause(WhereClause whereClause, QueryModel queryModel, Int32 index)
   at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   at NHibernate.Linq.Visitors.QueryModelVisitor.Visit()
   at NHibernate.Linq.Visitors.QueryModelVisitor.GenerateHqlQuery(QueryModel queryModel, VisitorParameters parameters, Boolean root, Nullable`1 rootReturnType)
   at NHibernate.Linq.NhLinqExpression.Translate(ISessionFactoryImplementor sessionFactory, Boolean filter)
   at NHibernate.Hql.Ast.ANTLR.ASTQueryTranslatorFactory.CreateQueryTranslators(IQueryExpression queryExpression, String collectionRole, Boolean shallow, IDictionary`2 filters, ISessionFactoryImplementor factory)
   at NHibernate.Engine.Query.QueryPlanCache.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow, IDictionary`2 enabledFilters)
   at NHibernate.Impl.AbstractSessionImpl.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow)
   at NHibernate.Impl.AbstractSessionImpl.CreateQuery(IQueryExpression queryExpression)
   at NHibernate.Linq.DefaultQueryProvider.PrepareQuery(Expression expression, IQuery& query)
   at NHibernate.Linq.DefaultQueryProvider.Execute(Expression expression)
   at NHibernate.Linq.DefaultQueryProvider.Execute[TResult](Expression expression)
```

### PR DESCRIPTION
**Pull request by fredericDelaporte**
fix #2937

Replace #2938.

### CODE REVIEW

File: src/NHibernate.Test/Async/Hql/EntityJoinHqlTest.cs [hunk 1]
```diff
@@ -346,9 +346,9 @@ public async Task NullableEntityProjectionAsync()
 				var withNullManyToOneList = await (session.Query<NullableOwner>().Where(x => x.ManyToOne == null).ToListAsync());
 				var withNullManyToOneJoinedList =
 					await ((from x in session.Query<NullableOwner>()
-							from x2 in session.Query<NullableOwner>()
-							where x == x2 && x.ManyToOne == null && x.OneToOne.Name == null
-							select x2).ToListAsync());
+					from x2 in session.Query<NullableOwner>()
+					where x == x2 && x.ManyToOne == null && x.OneToOne.Name == null
+					select x2).ToListAsync());
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/Async/Hql/EntityJoinHqlTest.cs:351]: I am not sure why those changes occur. Let get rid of them, if that is not an unstable generation artifact.

File: src/NHibernate.Test/Linq/TryGetMappedTests.cs [hunk 2]
```diff
@@ -807,7 +819,7 @@ public void NotNhQueryableTest()
 			Assert.That(() => expectedMemberType(memberType), $"Invalid member type: {memberType?.Name ?? "null"}");
 			Assert.That(() => expectedComponentType(componentType), $"Invalid component type: {componentType?.Name ?? "null"}");
 
-			if (found)
+			if (found && (componentType == null || componentType.PropertyNullability != null))
```
Comment (bahusoid) [inline src/NHibernate.Test/Linq/TryGetMappedTests.cs:822]: Maybe instead make nullability parameter nullable. And check that `_tryGetMappedNullability` returns false when it's null.

File: src/NHibernate.Test/Async/Hql/EntityJoinHqlTest.cs (hunk 1 above)
Comment (bahusoid) [reply src/NHibernate.Test/Async/Hql/EntityJoinHqlTest.cs:351]: I think I just made some whitespace changes for #2772 and forgot to regenerate async code.

File: src/NHibernate.Test/Linq/TryGetMappedTests.cs (hunk 2 above)
Comment (fredericDelaporte) [reply src/NHibernate.Test/Linq/TryGetMappedTests.cs:822]: I adding a commit doing what I think you are meaning, but I do not see this as an improvement. It looks to me as a less understandable way of testing, and maybe more brittle, subject to cause the test to fail if we start implementing that nullability property for custom composite type.

Comment (bahusoid) [reply src/NHibernate.Test/Linq/TryGetMappedTests.cs]: But with your initial change `TryGetMappedNullability` call is not tested at all for problematic case. Isn't it?  So your changes were not covered.

> subject to cause the test to fail if we start implementing that nullability property for custom composite type.

I see no problem with adjusting existing test cases when behavior is changed.

nit: I would also add optional `nullability` parameter to `AssertResult` method and keep `AssertResult` call it test case

Comment (fredericDelaporte) [reply src/NHibernate.Test/Linq/TryGetMappedTests.cs]: Ok, I thought the test parsing the query would also cause the method to be called, but I had not checked. I have added your nit.

**Review by bahusoid — APPROVED** [review]

**Pull request merged**